### PR TITLE
Take margin into account on max dimension

### LINKED ISSUE #466: maxHeight incorrectly includes margins in the calculation
When a node has both a max height and a vertical margin, it appears that Yoga incorrectly considers the margins to take up part of the max height.

### Repro Steps

Here's some HTML representing a case that repros the bug:

```html
<div style="width: 250px; height: 250px;">
  <div style="width: 100px; height: 100px; max-height: 100px; margin-top: 20px;"></div>
</div>
```

Here's the equivalent C code:

```c
  const YGNodeRef root = YGNodeNew();
  YGNodeStyleSetFlexDirection(root, YGFlexDirectionRow);
  YGNodeStyleSetWidth(root, 250);
  YGNodeStyleSetHeight(root, 250);
  
  const YGNodeRef root_child0 = YGNodeNew();
  YGNodeStyleSetWidth(root_child0, 100);
  YGNodeStyleSetHeight(root_child0, 100);
  YGNodeStyleSetMaxHeight(root_child0, 100);
  YGNodeStyleSetMargin(root_child0, YGEdgeTop, 20);
  YGNodeInsertChild(root, root_child0, 0);
  
  YGNodeCalculateLayout(root, YGUndefined, YGUndefined, YGDirectionLTR);
  YGNodePrint(root, YGPrintOptionsLayout | YGPrintOptionsChildren);
```

### Expected Output

Child's height is 100:

```
{layout: {width: 250, height: 250, top: 0, left: 0}, children: [
  {layout: {width: 100, height: 100, top: 20, left: 0}, },
]},
```

This JSFiddle illustrates the expected behavior: https://jsfiddle.net/4mr6tbqq/

### Actual Output

Child's height is 80:

```
{layout: {width: 250, height: 250, top: 0, left: 0}, children: [
  {layout: {width: 100, height: 80, top: 20, left: 0}, },
]},
```

### PR DESCRIPTION
**Pull request by woehrl01**
We need to take the margin into account if we clip on max dimension. Fixes #466.

### CODE REVIEW

File: yoga/Yoga.c [hunk 1]
```diff
@@ -2443,16 +2446,18 @@ static void YGNodelayoutImpl(const YGNodeRef node,
...
-            &childMainMeasureMode,
-            &childMainSize);
-        YGConstrainMaxSizeForMode(
-            YGValueResolve(&currentRelativeChild->style.maxDimensions[dim[crossAxis]],
-                           availableInnerHeight),
-            &childCrossMeasureMode,
-            &childCrossSize);
+        YGConstrainMaxSizeForMode(currentRelativeChild,
+                                  mainAxis,
+                                  availableInnerWidth,
```
Comment (rigdern) [inline yoga/Yoga.c]: This this be the main width rather than `availableInnerWidth`?

File: yoga/Yoga.c [hunk 2]
```diff
@@ -2443,16 +2446,18 @@ static void YGNodelayoutImpl(const YGNodeRef node,
...
+                                  availableInnerWidth,
+                                  &childMainMeasureMode,
+                                  &childMainSize);
+        YGConstrainMaxSizeForMode(currentRelativeChild,
+                                  crossAxis,
+                                  availableInnerHeight,
```
Comment (rigdern) [inline yoga/Yoga.c]: Should this be the cross axis size rather than `availableInnerHeight`?

File: yoga/Yoga.c [hunk 3]
```diff
@@ -1480,14 +1487,10 @@ static void YGNodeComputeFlexBasisForChild(const YGNodeRef node,
...
-    YGConstrainMaxSizeForMode(YGValueResolve(&child->style.maxDimensions[YGDimensionWidth],
-                                             parentWidth),
-                              &childWidthMeasureMode,
-                              &childWidth);
-    YGConstrainMaxSizeForMode(YGValueResolve(&child->style.maxDimensions[YGDimensionHeight],
-                                             parentHeight),
-                              &childHeightMeasureMode,
-                              &childHeight);
+    YGConstrainMaxSizeForMode(
+        child, YGDimensionWidth, parentWidth, parentWidth, &childWidthMeasureMode, &childWidth);
```
Comment (rigdern) [inline yoga/Yoga.c]: Should `YGDimensionWidth` and `YGDimensionHeight` be changed to be `YGFlexDirection*` enums instead of `YGDimension*` enums?

**Comment by rigdern** [comment]
@woehrl01 I noticed some potential mistakes when trying to get this change working with React Native 0.42. I left comments on the commit.

**Comment by woehrl01** [comment]
@rigdern You're absolutely right! Thanks a lot for having a look!

**Pull request closed without merging**